vault-plugins/flant_iam_auth/backend/path_tenant.go

```diff
@@ -129,15 +135,25 @@ func (b *flantIamAuthBackend) listProjects(ctx context.Context, req *logical.Req
 		return nil, err
 	}
 
-	result := make([]ext.SafeProject, 0, len(projects))
+	acceptedTenants, acceptedProjects, err := b.availableTenantsAndProjectsByEntityIDOwner(ctx, req)
+	if err != nil {
+		return responseErrMessage(req, fmt.Sprintf("collect acceptedTenants & acceptedProjects: %s", err.Error()),
+			http.StatusInternalServerError)
+	}
+
+	result := make([]model.SafeProject, 0, len(projects))
 
 	for _, project := range projects {
-		res := ext.SafeProject{
-			UUID:       project.UUID,
-			TenantUUID: project.TenantUUID,
-			Version:    project.Version,
+		_, tenantAcccepted := acceptedTenants[project.TenantUUID]
+		_, projectAcccepted := acceptedProjects[project.UUID]
+		if tenantAcccepted && projectAcccepted {
+			res := model.SafeProject{
+				UUID:       project.UUID,
+				TenantUUID: project.TenantUUID,
+				Version:    project.Version,
+			}
+			result = append(result, res)
 		}
-		result = append(result, res)
 	}
```

Comment:
I'd say the code is somewhat messy here. The vault framework detail dives too deeply. This transport layer does too much from its behalf:

1. takes tx
2. calls availableTenantsAndProjectsByEntityIDOwner(ctx, req)
  3. also takes tx
  4. calls revealEntityIDOwner(ctx, req)
    5. takes tx 
    6. context is not used
    7. req is not used, only EntityID is of use
    6. calls IdentityAPI which is not injected

I would expect lines 133–156 to be a usecase one-line call. NewIdentityAPI could be instantiated here and passed to the usecase as dependency

I would suggest making one more iteration on that

Reply:
Probably I have done you suggested